### PR TITLE
Improving multiline context snippet

### PR DESCRIPTION
**Pull request by eddynaka**
## Change
- this will improve contextSnippet when we have only one line containing everything

### CODE REVIEW

File: src/Sarif/FileRegionsCache.cs [hunk 1]
```diff
@@ -156,15 +156,42 @@ internal Region ConstructMultilineContextSnippet(Region inputRegion, Uri uri)
...
 
+            const int maxSnippetLength = 512;
+            const int reasonableSnippetLength = 256;
+            const int smallSnippetLength = 128;
+
+            // If this is big enough, we don't need to create a new snippet.
+            if (inputRegion.Snippet?.Text.Length >= reasonableSnippetLength)
+            {
+                return null;
+            }
```
Comment (michaelcfanning) [inline src/Sarif/FileRegionsCache.cs:167]: shouldn't we actually be checking the region size, not the snippet size? basically, if a region exceeds a certain size, no context region is required. so for this case, the region should have the snippet. your code will permit the following case:

region is 10k in size but has no snippet
we then create a context region + snippet which is smaller than the region! this breaks the SARIF spec. :)

File: src/Sarif/FileRegionsCache.cs [hunk 2]
```diff
@@ -156,15 +156,42 @@ internal Region ConstructMultilineContextSnippet(Region inputRegion, Uri uri)
                 return null;
             }
 
+            const int maxSnippetLength = 512;
+            const int reasonableSnippetLength = 256;
+            const int smallSnippetLength = 128;
+
+            // If this is big enough, we don't need to create a new snippet.
+            if (inputRegion.CharLength >= reasonableSnippetLength)
```
[the remark this answers is not in the document]
Comment (eddynaka) [reply src/Sarif/FileRegionsCache.cs]: Ok, so, we will just check this value if charOffset != -1.

---
In reply to: [578571743](https://github.com/microsoft/sarif-sdk/pull/2288#discussion_r578571743) [](ancestors = 578571743)

File: src/Sarif/FileRegionsCache.cs (hunk 1 above)
Comment (eddynaka) [reply src/Sarif/FileRegionsCache.cs:167]: changed to the charLength.

---
In reply to: [578048420](https://github.com/microsoft/sarif-sdk/pull/2288#discussion_r578048420) [](ancestors = 578048420)

File: src/Sarif/FileRegionsCache.cs [hunk 3]
```diff
@@ -156,15 +156,42 @@ internal Region ConstructMultilineContextSnippet(Region inputRegion, Uri uri)
...
+            // We are going to grab 128 characters from left and from right if possible.
+            region.CharOffset = inputRegion.CharOffset < smallSnippetLength
+                ? 0
+                : inputRegion.CharOffset - smallSnippetLength;
+
+            region.CharLength = inputRegion.CharLength + inputRegion.CharOffset + smallSnippetLength < newLineIndex.Text.Length
+                ? inputRegion.CharLength + inputRegion.CharOffset + smallSnippetLength
+                : newLineIndex.Text.Length - region.CharOffset;
+
             return this.PopulateTextRegionProperties(region, uri, populateSnippet: true);
```
[the remark this answers is not in the document]
Comment (eddynaka) [reply src/Sarif/FileRegionsCache.cs]: Added the debug.Assert to verify if the new length >= the original

---
In reply to: [578572270](https://github.com/microsoft/sarif-sdk/pull/2288#discussion_r578572270) [](ancestors = 578572270)

File: src/Sarif/FileRegionsCache.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (michaelcfanning) [reply src/Sarif/FileRegionsCache.cs]: Wait, this isn't right, is it? If CharOffset == -1, this method can't function, can it?

---
In reply to: [578583524](https://github.com/microsoft/sarif-sdk/pull/2288#discussion_r578583524) [](ancestors = 578583524,578571743)

[the remark this answers is not in the document]
Comment (eddynaka) [reply src/Sarif/FileRegionsCache.cs]: if == -1, then, ur region is using startline / startcolumn, with that, the populateTextRegion handle both cases.

---
In reply to: [578586111](https://github.com/microsoft/sarif-sdk/pull/2288#discussion_r578586111) [](ancestors = 578586111,578583524,578571743)

**Review by michaelcfanning — APPROVED** [review]
:shipit:

**Pull request merged**